Terraform policy: build plugins for correct OS/arch (#2966)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/write/plugins.mdx

# Write policy plugins

Plugins provide a way to extend Terraform policy with additional functions that you can use to customize your policy authoring experience. Write Terraform policy plugins in Golang using the [`terraform-policy-plugin-framework`](https://github.com/hashicorp/terraform-policy-plugin-framework) open source library, and compile them for your testing environment and HCP Terraform execution environment.

@include "beta.mdx"

## Requirements

> You must compile Terraform policy plugins and include them in your policy VCS repository before you can use them in your policies. You must compile plugins for an architecture that matches the environment in which Terraform policy will be run.

Reviewer: ```suggestion
You must compile Terraform policy plugins and include them in your policy VCS repository before you can use them in your policies. You must compile plugins for an architecture that matches the environment in which Terraform policy runs.
```
style nit: use present tense

</CodeBlockConfig>

> After you implement your plugin in Go, compile it into a binary for your target platform and store both the source and binary in the same VCS repository as your policy files. Because plugins are natively compiled Go binaries, you must compile them for the specific OS and architecture where your policy set will run.

Reviewer 2: Minor: consider clarifying that only the binary matching the execution environment belongs in the repo — with local testing on macOS in the picture (see comment on examples.mdx), people may otherwise commit their darwin test build alongside the Linux one.

> After you implement your plugin in Go, compile it into a binary for your target platform and store both the source and binary in the same VCS repository as your policy files. Because plugins are natively compiled Go binaries, you must compile them for the specific OS and architecture where your policy set will run.

Reviewer: ```suggestion
After you implement your plugin in Go, compile it into a binary for your target platform and store both the source and binary in the same VCS repository as your policy files. Because plugins are natively compiled Go binaries, you must compile them for the specific OS and architecture where your policy set runs.
```
style nit

└── README.md
```

> When you enforce your policies in HCP Terraform, your VCS repository must also contain the plugin binary, compiled for the OS and architecture of the system where the plugin will run.

Reviewer: ```suggestion
When you enforce your policies in HCP Terraform, your VCS repository must also contain the plugin binary, compiled for the OS and architecture of the system where the plugin runs.
```
style nit... or "where you plan to run the plugin"

## Compile a plugin

> Before you can use your plugin in your policies, you must compile your plugin into a binary that is compatible with the target architecture. When you test your policies on your local machine, you compile the plugin to match your system's OS and architecture. When you deploy your policies to HCP Terraform, you must compile the plugin to match the OS and architecture HCP Terraform uses to evaluate your policies: either an HCP Terraform worker or an HCP Terraform agent that you have configured for your organization.

Reviewer 3: ```suggestion
Before you can use a plugin in your policies, you must compile the plugin into a binary that is compatible with the target architecture. When testing policies on your local machine, compile the plugin to match your system's OS and architecture. 

When deploying policies to HCP Terraform, you must compile the plugin to match the OS and architecture HCP Terraform uses to evaluate your policies. When the HCP Terraform execution mode is set to **Remote**, the plugin must match the HCP Terraform worker. When set to **Agent**, the plugin must match the HCP Terraform agent that you have configured for your organization. Refer to [Default execution mode](/terraform/cloud-docs/users-teams-organizations/organizations/settings#default-execution-mode) in the HCP Terraform documentation for more information.
```
Tried to edit down some of the "you"s and "your"s because it felt like fingers pointing at me. 

I think this is what we mean about matching the HCP TF architecture, right?

> Before you can use your plugin in your policies, you will compile it. You must compile your plugin into a binary that is compatible with the architecture it will run on. When you test your policies on your local machine, you will compile the plugin to match your system's OS and architecture. When you deploy your policies to HCP Terraform, you will compile the plugin to match the OS and architecture HCP Terraform will use to evaluate your policies: either an HCP Terraform worker or an HCP Terraform agent that you have configured for your organization.

Reviewer: ```suggestion
Before you can use your plugin in your policies, you must compile your plugin into a binary that is compatible with the target architecture. When you test your policies on your local machine, you compile the plugin to match your system's OS and architecture. When you deploy your policies to HCP Terraform, you must compile the plugin to match the OS and architecture HCP Terraform uses to evaluate your policies: either an HCP Terraform worker or an HCP Terraform agent that you have configured for your organization.
```

> HCP Terraform workers run a `linux/amd64` architecture. If you use HCP Terraform agents, your agents may use either a `linux/amd64` or a `linux/arm64` environment. Refer to [VCS setup for Terraform policy](/terraform/cloud-docs/agents/requirements#environment) for more information about evaluating your policies in HCP Terraform.

Reviewer 2: The link text says "VCS setup for Terraform policy" but the URL points to the agent requirements page. Suggest changing the text to something like "HCP Terraform agent requirements".

Author: Updated the link; we want to point to the HCP Terraform policy set docs for this.

## Compile a plugin

> Before you can use your plugin in your policies, you must compile it. Plugins are natively compiled Go binaries that Terraform policy runs within your HCP Terraform execution environment. HCP Terraform worker VMs run a `linux/amd64` architecture. If you use HCP Terraform agents, your agents may use either a `linux/amd64` or a `linux/arm64` environment. Refer to [Agent requirements](/terraform/cloud-docs/agents/requirements#environment) for more information.

Reviewer 2: Worth telling self-hosted agent users *how* to check, since they control the image: run `uname -m` on the agent and map the output to Go's names — `x86_64` → `GOARCH=amd64`, `aarch64` → `GOARCH=arm64`. The raw `uname -m` values don't match the `GOARCH` values, and that mismap is an easy mistake.

$ go mod init example
```

Replace `example` with the name of your plugin.

Download and verify go modules.

```shell-session
$ go mod tidy
```

Build the plugin for your target platform.

To test your plugin on your local machine with the `tfpolicy test` command, build it with the following command:

```shell-session
$ go build -o ../../bin/example main.go
```

> Do not check this binary into your VCS repository, since it only runs on a system compatible with your local machine.

Reviewer 3: ```suggestion
Do not check this binary into your VCS repository because it only runs on a system compatible with your local machine.
```

> Do not check this binary into your VCS repository, since it will only run on a system compatible with your local machine.

Reviewer: ```suggestion
Do not check this binary into your VCS repository, since it only runs on a system compatible with your local machine.
```

To build the plugin for HCP Terraform workers or agents running `linux/amd64`, run the following command:

```shell-session
$ CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o ../../bin/example main.go
```

> To build the plugin for HCP Terraform agents running `linux/arm64` agents, run the following command:

Reviewer 2: Duplicated word: "agents running `linux/arm64` agents" - drop the trailing "agents".

```shell-session
$ CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -o ../../bin/example main.go
```

> The `CGO_ENABLED=0` environment variable creates a portable, statically linked binary. The `GOOS` and `GOARCH` environment variables specify the target operating system and architecture. Be sure to commit the correct binary for your HCP Terraform envirotnment to your version control repository.

Reviewer 2: Typo: "envirotnment" should be "environment".

$ go mod tidy
```

> Build the plugin for your target platform. For HCP Terraform workers or agents using `linux/amd64`, run:

Reviewer 4: ```suggestion
Build the plugin for your target platform. Run the following command to build the plugin for HCP Terraform workers or agents using `linux/amd64`:
```

```shell-session
$ CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o ../../bin/example main.go
```

> For agents using `linux/arm64`, run:

Reviewer 4: ```suggestion
Run the following command for `linux/arm64` agents:
```

```shell-session

> $ CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -o ../../bin/example main.go

Reviewer 2: The docs don't state the mixed-fleet limitation: the `plugins` block points at a single binary, so one policy set cannot serve agent pools that span both `amd64` and `arm64` — there is no per-architecture selection. Users with mixed pools should standardize the pool that runs this policy set on one architecture. Better to state this explicitly than let users discover it via handshake failures.

```

> The `CGO_ENABLED=0` flag ensures the binary is statically linked and portable. The `GOOS` and `GOARCH` environment variables specify the target operating system and architecture.

Reviewer 4: ```suggestion
The `CGO_ENABLED=0` flag creates a portable, statically linked binary. The `GOOS` and `GOARCH` environment variables specify the target operating system and architecture.
```

> The `CGO_ENABLED=0` flag creates a portable, statically linked binary. The `GOOS` and `GOARCH` environment variables specify the target operating system and architecture.

Reviewer 2: Two additions here would save support round-trips:

1. A verification step so authors can confirm what they built before shipping it:

```shell-session
$ file ../../bin/example
```

should report the target arch, e.g. `ELF 64-bit ... x86-64` or `aarch64`.

2. What a mismatch looks like: if the binary doesn't match the agent's architecture, policy evaluation fails with a plugin handshake/connection error — not an explicit "wrong architecture" message. Naming that symptom here lets users connect the error back to this doc.

Author: Do you happen to have the exact text of the error message? If not, I can reproduce, but that will also help people find it when they do run into it.

Author: I reproduced the error and included the text in hopes of helping with SEO of people searching for it.

</CodeBlockConfig>

> The value of the `source` attribute must be a string that is the relative path to a binary in the same VCS repository as your policies. The binary must be a valid Terraform policy Go plugin server. Use the [`terraform-policy-plugin-framework`](https://github.com/hashicorp/terraform-policy-plugin-framework) open source library to ensure your plugin provides the correct interface. Be sure to commit a binary compiled for the correct architecture to your VCS repository. Refer to [VCS setup for Terraform policy](/terraform/cloud-docs/workspaces/policy-enforcement/terraform-policy-vcs) for more information.

Reviewer 2: This link is missing the `manage-policy-sets` segment, so it 404s. It should be `/terraform/cloud-docs/workspaces/policy-enforcement/manage-policy-sets/terraform-policy-vcs`.
